### PR TITLE
Returning stderr in invocation failure errors

### LINKED ISSUE #820: RPC targeting a non-existing component fails with no useful error message
For example:

```Error: InternalError: GolemErrorBody { golem_error: PreviousInvocationFailed(GolemErrorPreviousInvocationFailed { details: "error while executing at wasm backtrace:
    0: 0xbb41ae - <unknown>!<wasm function 110>
    1: 0xbb504a - <unknown>!<wasm function 119>
    2: 0xbb48e7 - <unknown>!<wasm function 118>
    3: 0xbb484e - <unknown>!<wasm function 117>
    4: 0xbb0ea3 - <unknown>!<wasm function 65>
    5: 0xbb0be9 - <unknown>!<wasm function 59>
    6: 0xbae37d - <unknown>!<wasm function 39>
    7: 0xbb01c4 - <unknown>!<wasm function 52>
    8:  0x92d - <unknown>!<wasm function 18>
    9: 0xba4de5 - wit-component:shim!indirect-demo:archive-stub/stub-archive-[method]api.store
   10: 0x776c30 - <unknown>!<wasm function 10240>
   11: 0x34cfcf - <unknown>!<wasm function 603>
   12: 0xa9c98 - <unknown>!<wasm function 158>
   13: 0x3372b4 - <unknown>!<wasm function 571>
   14: 0x34ce94 - <unknown>!<wasm function 603>
   15: 0x693cff - <unknown>!<wasm function 4813>
   16: 0x50497a - <unknown>!<wasm function 1819>
   17: 0x3b91b8 - <unknown>!<wasm function 795>
   18: 0x78eb94 - <unknown>!demo:lst/api#archive

Caused by:
    wasm trap: wasm `unreachable` instruction executed

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: wasmtime::runtime::trap::from_runtime_box
   2: wasmtime::runtime::func::invoke_wasm_and_catch_traps
   3: wasmtime::runtime::component::func::Func::call_impl
   4: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
   5: wasmtime_fiber::unix::fiber_start
   6: wasmtime_fiber_start_21_0_1
   7: wasmtime_fiber::Fiber<Resume,Yield,Return>::resume
   8: wasmtime::runtime::store::<impl wasmtime::runtime::store::context::StoreContextMut<T>>::on_fiber::{{closure}}::FiberFuture::resume
   9: <wasmtime::runtime::store::<impl wasmtime::runtime::store::context::StoreContextMut<T>>::on_fiber::{{closure}}::FiberFuture as core::future::future::Future>::poll
  10: golem_worker_executor_base::invocation::invoke::{{closure}}
  11: golem_worker_executor_base::invocation::invoke_or_fail::{{closure}}
  12: golem_worker_executor_base::invocation::invoke_worker::{{closure}}
  13: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  14: <golem_worker_executor_base::durable_host::DurableWorkerCtx<Ctx> as golem_worker_executor_base::workerctx::ExternalOperations<Ctx>>::prepare_instance::{{closure}}
  15: <cloud_worker_executor::context::Context as golem_worker_executor_base::workerctx::ExternalOperations<cloud_worker_executor::context::Context>>::prepare_instance::{{closure}}
  16: golem_worker_executor_base::worker::RunningWorker::invocation_loop::{{closure}}
  17: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  18: golem_worker_executor_base::worker::RunningWorker::new::{{closure}}
  19: tokio::runtime::task::core::Core<T,S>::poll
  20: tokio::runtime::task::harness::Harness<T,S>::poll
  21: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
  22: tokio::runtime::scheduler::multi_thread::worker::Context::run
  23: tokio::runtime::context::scoped::Scoped<T>::set
  24: tokio::runtime::context::runtime::enter_runtime
  25: tokio::runtime::scheduler::multi_thread::worker::run
  26: tokio::runtime::task::core::Core<T,S>::poll
  27: tokio::runtime::task::harness::Harness<T,S>::poll
  28: tokio::runtime::blocking::pool::Inner::run
  29: std::sys_common::backtrace::__rust_begin_short_backtrace
  30: core::ops::function::FnOnce::call_once{{vtable.shim}}
  31: std::sys::pal::unix::thread::Thread::new::thread_start
  32: <unknown>
  33: <unknown>" }) }```

### PR DESCRIPTION
**Pull request by vigoo**
Resolves #820 

Always returns the worker's `stderr` with invocation failures. The size is limited by the worker executor's event buffer size. 

The oplog is not affected by this change - for running workers, the stderr entries are collected after writing the error to the oplog from the worker's event buffer; when recovering invocation results from the oplog, we are looking back to read `Log` entries (this could be further optimized to do batched reads but it's not implemented yet).

Also partially implements #774 (the invocation events are there in the executor and through gRPC, but not added to the websocket and not used yet in the CLI)

### CODE REVIEW

File: golem-common/src/model/oplog.rs [hunk 1]
```diff
@@ -674,13 +674,30 @@ pub enum WorkerError {
...
-            WorkerError::StackOverflow => write!(f, "Stack overflow"),
-            WorkerError::OutOfMemory => write!(f, "Out of memory"),
+            WorkerError::Unknown(message) => format!("WE UNKNOWN {message}{error_logs}"),
+            WorkerError::InvalidRequest(message) => format!("WE IR {message}{error_logs}"),
+            WorkerError::StackOverflow => format!("Stack overflow{error_logs}"),
+            WorkerError::OutOfMemory => format!("Out of memory{error_logs}"),
         }
     }
 }
+
```
Comment (noise64) [inline golem-common/src/model/oplog.rs:692]: is this needed?

File: golem-worker-executor-base/src/durable_host/mod.rs [hunk 2]
```diff
@@ -1283,9 +1307,11 @@ impl<Ctx: WorkerCtx + DurableWorkerCtxView<Ctx>> ExternalOperations<Ctx> for Dur
                     .unwrap_or(default_retry_config),
                 &last_error,
             );
-            if let Some(last_error) = last_error {
-                debug!("Recovery decision after {last_error}: {decision:?}");
-            }
+
+            // TODO
```
Comment (noise64) [inline golem-worker-executor-base/src/durable_host/mod.rs:1311]: is this todo or should be dropped?

File: golem-worker-executor-base/src/error.rs [hunk 3]
```diff
@@ -342,7 +342,7 @@ impl From<anyhow::Error> for GolemError {
     fn from(error: anyhow::Error) -> Self {
         match error.root_cause().downcast_ref::<InterruptKind>() {
             Some(kind) => GolemError::Interrupted { kind: kind.clone() },
-            None => GolemError::runtime(format!("{error}")),
+            None => GolemError::runtime(format!("KACSA {error:#?}")),
```
Comment (noise64) [inline golem-worker-executor-base/src/error.rs:345]: 🦆

File: golem-worker-executor-base/src/invocation.rs [hunk 4]
```diff
@@ -468,10 +496,20 @@ async fn call_exported_function<Ctx: FuelManagement + Send>(
...
         );
     }
+
+    if let Some(idempotency_key) = idempotency_key {
+        store
+            .data()
+            .get_public_state()
+            .event_service()
+            .emit_invocation_finished(raw_function_name, &idempotency_key, is_live);
+        // TODO: reevaluate is_live
```
Comment (noise64) [inline golem-worker-executor-base/src/invocation.rs:510]: is this needed?

File: golem-worker-executor-base/src/model.rs [hunk 5]
```diff
@@ -220,22 +220,24 @@ impl TrapType {
                             Some(GolemError::InvalidRequest { details }) => {
                                 TrapType::Error(WorkerError::InvalidRequest(details.clone()))
                             }
-                            _ => TrapType::Error(WorkerError::Unknown(format!("{:?}", error))),
+                            _ => {
+                                TrapType::Error(WorkerError::Unknown(format!("POCOK {:#?}", error)))
```
Comment (noise64) [inline golem-worker-executor-base/src/model.rs:224]: 🐹

File: golem-worker-executor-base/src/worker.rs [hunk 6]
```diff
@@ -780,25 +785,41 @@ impl<Ctx: WorkerCtx> Worker<Ctx> {
...
                 InvocationResult::Cached {
-                    result: Err(TrapType::Error(error)),
+                    result:
+                        Err(FailedInvocationResult {
+                            trap_type: TrapType::Error(error),
+                            stderr,
+                        }),
                     ..
-                } => LookupResult::Complete(Err(GolemError::runtime(error.to_string()))),
+                } => LookupResult::Complete(Err(GolemError::runtime(error.to_string(&stderr)))), // TODO
```
Comment (noise64) [inline golem-worker-executor-base/src/worker.rs:816]: todo

File: golem-worker-executor-base/src/durable_host/mod.rs [hunk 7]
```diff
@@ -1366,6 +1406,46 @@ async fn last_error_and_retry_count<T: HasOplogService>(
...
+pub(crate) async fn recover_stderr_logs<T: HasOplogService + HasConfig>(
+    this: &T,
+    owned_worker_id: &OwnedWorkerId,
+    last_oplog_idx: OplogIndex,
+) -> String {
+    let max_count = this.config().limits.event_history_size;
+    let mut idx = last_oplog_idx;
+    let mut stderr_entries = Vec::new();
+    loop {
+        // TODO: this could be read in batches to speed up the process
```
Comment (justcoon) [inline golem-worker-executor-base/src/durable_host/mod.rs:1420]: is this TODO for future improvement ?

**Review by justcoon — APPROVED** [review]

**Pull request merged**